[AArch64] Fix handling of x29/x30 in inline assembly clobbers

lalinsky opened the pull request:
The AArch64 backend was silently ignoring inline assembly clobbers when
numeric register names (x29, x30) were used instead of their
architectural aliases (fp, lr). I found this bug via inline assembly
in Zig, which not normalize the register names the way clang does.

There is an incoplete workaround for this in Rust, but that only
handles `x30/lr`, not `x29/fp`. I thought it would make
sense to fix this properly rather than adding a workaround to Zig.

This patch adds explicit handling in getRegForInlineAsmConstraint() to
map both numeric and alias forms to the correct physical registers,
following the same pattern used by the RISC-V backend.

I've left `x31/sp` without changes, it would nice to have to have
warning when trying to clobber `x31`, just like there is for `sp`,
but that register needs different handling, so it's best done
separately.

If you have code like this:

    define void @clobber_x30() nounwind {
      tail call void asm sideeffect "nop", "~{x30}"()
      ret void
    }

Here is the generated assembly before:

    clobber_x30:                            // @clobber_x30
            //APP
            nop
            //NO_APP
            ret

And after:

    clobber_x30:                            // @clobber_x30
            str     x30, [sp, #-16]!                // 8-byte Folded Spill
            //APP
            nop
            //NO_APP
            ldr     x30, [sp], #16                  // 8-byte Folded Reload
            ret

lalinsky committed:
[AArch64] Fix handling of x29/x30 in inline assembly clobbers

The AArch64 backend was silently ignoring inline assembly clobbers when
numeric register names (x29, x30) were used instead of their
architectural aliases (fp, lr). I found this bug via inline assembly
in Zig, which not normalize the register names the way clang does.

There is an incoplete workaround for this in Rust, but that only
handles `x30/lr`, not `x29/fp`. I thought it would make
sense to fix this properly rather than adding a workaround to Zig.

This patch adds explicit handling in getRegForInlineAsmConstraint() to
map both numeric and alias forms to the correct physical registers,
following the same pattern used by the RISC-V backend.

I've left `x31/sp` without changes, it would nice to have to have
warning when trying to clobber `x31`, just like there is for `sp`,
but that register needs different handling, so it's best done
separately.

If you have code like this:

    define void @clobber_x30() nounwind {
      tail call void asm sideeffect "nop", "~{x30}"()
      ret void
    }

Here is the generated assembly before:

    clobber_x30:                            // @clobber_x30
            //APP
            nop
            //NO_APP
            ret

And after:

    clobber_x30:                            // @clobber_x30
            str     x30, [sp, #-16]!                // 8-byte Folded Spill
            //APP
            nop
            //NO_APP
            ldr     x30, [sp], #16                  // 8-byte Folded Reload
            ret

DavidSpickett commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -13126,6 +13127,19 @@ AArch64TargetLowering::getRegForInlineAsmConstraint(
+                               .Default(AArch64::NoRegister);
+  if (XRegFromAlias != AArch64::NoRegister) {
+    return std::make_pair(XRegFromAlias, &AArch64::GPR64RegClass);
+  }
```
https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

lalinsky commented:
The CI failures are interesting, they are actually testing the incorrect behavior:

```
  tail call void asm sideeffect "nop", "~{x0},~{x1},~{x2},~{x3},~{x4},~{x5},~{x6},~{x7},~{x8},~{x9},~{x10},~{x11},~{x12},~{x13},~{x14},~{x15},~{x16},~{x17},~{x18},~{x19},~{x20},~{x21},~{x22},~{x23},~{x24},~{x25},~{x26},~{x27},~{x28},~{x29},~{x30},~{x31}"() nounwind
```

DavidSpickett commented:
> https://github.com/rust-lang/rust/blob/21cbbdc44de84e3ea99bca239091e5d1c49af654/compiler/rustc_codegen_llvm/src/asm.rs#L487-L490

Could you open an issue in Rust for this? I don't know how they handle backports/updates/cherry-picks but I'm sure someone over there can do the right thing for it if you open an issue.

lalinsky commented on llvm/test/CodeGen/AArch64/inline-asm-clobber-x29-x30.ll:
```diff
@@ -0,0 +1,36 @@
+; CHECK:       ldr x30, [sp
+  tail call void asm sideeffect "", "~{lr}"()
+  ret void
+}
```
Hm, I think sp should not have been added to the stringswitch, instead of:

```
warning: inline asm clobber list contains reserved registers: SP
```

it now prints:

```
error: register 'SP' allocated for constraint '{sp}' does not match required type
error: register 'SP' allocated for constraint '{x31}' does not match required type
```

Maybe I should remove it and not worry about it? It would be stupid trying to mark sp/x31 as clobbered anyway.

DavidSpickett reviewed (commented):
I was able to confirm the missing save/restore with https://godbolt.org/z/ETjrPc1nG 

Please add an example to the commit message showing the before/after.

I understand the change and the new test cases, but am not sure why each of the existing tests had to change. Can you explain?

I expected that the current use should still work, or if it didn't, all existing tests would remove x30 and x31, but some of them have x29 and x31 removed.

I'm missing something but regardless, please point it out for me :)

lalinsky commented:
> I understand the change and the new test cases, but am not sure why each of the existing tests had to change. Can you explain?
> 
> I expected that the current use should still work, or if it didn't, all existing tests would remove x30 and x31, but some of them have x29 and x31 removed.

fp/sp can't be really marked as clobbered, they trigger warnings, and those warnings failed the tests. they literally depended on the fact that x29/x30/x31 are completely ignored. there is even the one test I mentioned in an earlier comment, that tries to clobber everything but x30, likely to show that llvm can use x30 for its own purposes in such situatoins, but that previously worked just by the fact that all of them were ignored. after I changed the behavior of x29, so that it clobbers fp, the behavior of the tests changed.

I can try again, now that I removed x31 from the mapping, but it seems silly to have x31 in the tests and depend on it being ignored.

DavidSpickett commented on llvm/test/CodeGen/AArch64/inline-asm-clobber-x29-x30.ll:
```diff
@@ -0,0 +1,36 @@
+; CHECK-LABEL: clobber_x29:
+; CHECK:       str x29, [sp
+; CHECK:       ldr x29, [sp
+  tail call void asm sideeffect "", "~{x29}"()
```
I know these are marked tailcall so in theory we should never get load/stores for stack frame management.

However, if you put a nop in the asm statement, you could make this into:
CHECK: str
CHECK-NEXT: nop
CHECK-NEXT: ldr

And it would be bit more robust and immediately clear what the intended behaviour is.

DavidSpickett replied:
Well, you stated the test intent above (thank you for that too, it often gets forgotten), but it can be visually more clear.

DavidSpickett reviewed (commented):
I think this needs a release note in https://github.com/llvm/llvm-project/blob/main/llvm/docs/ReleaseNotes.md#changes-to-the-aarch64-backend.

I suggest:
* A bug was fixed that caused LLVM IR inline assembly clobbers of the `x29` and `x30` registers to be ignored when they were written using their `xN` names instead of the ABI names `FP` and `LR`. Note that LLVM IR produced by Clang always uses the ABI names, but other frontends may not. (#link-to-this-pr)

Feel free to copy paste that if you think it's correct. And more importantly, if you had never found this problem, and read this release note during a Zig update, what would you want to see there?

alexrp commented:
It would be nice if we could get this into 22.x along with https://github.com/llvm/llvm-project/pull/179740.

DavidSpickett commented:
Sorry, this got lost in switching email accounts. Looking at it now.

DavidSpickett reviewed (approved):
LGTM with the comment wording fixed.

I assume you'll need me to merge this for you.

folkertdev commented:
Quick ping here @lalinsky, this seems like it's basically done, just needs a merge conflict resolved and a small fix to a comment?

lalinsky commented:
> Quick ping here @lalinsky, this seems like it's basically done, just needs a merge conflict resolved and a small fix to a comment?

Thanks, will rebase it and update the comment ASAP

lalinsky commented:
@folkertdev fixed, CI needs approval
